dhall/src/Dhall/Eval.hs

```diff
@@ -132,873 +105,1198 @@ instance Monoid (VChunks a) where
+                    VHLam (Typed "optional" (VConst Type)) $ \optional ->
+                    VHLam (Typed "some" (a ~> optional)) $ \some ->
+                    VHLam (Typed "none" optional) $ \none ->
+                    some `vApp` t
+                opt ->
+                    VHLam (OptionalFoldCl opt) $ \o ->
+                    VPrim $ \s ->
+                    VPrim $ \n ->
+                    VOptionalFold a opt o s n
+        OptionalBuild ->
+            VPrim $ \ ~a ->
+            VPrim $ \case
+                VHLam (OptionalFoldCl x) _ -> x
+                VPrimVar -> VOptionalBuild a VPrimVar
+                t ->       t
+                    `vApp` VOptional a
+                    `vApp` VHLam (Typed "a" a) VSome
+                    `vApp` VNone a
+        Record kts ->
+            VRecord (Map.sort (fmap evalE kts))
+        RecordLit kts ->
+            VRecordLit (Map.sort (fmap evalE kts))
+        Union kts ->
+            VUnion (Map.sort (fmap (fmap evalE) kts))
+        Combine t u ->
+            vCombine (evalE t) (evalE u)
+        CombineTypes t u ->
+            vCombineTypes (evalE t) (evalE u)
+        Prefer t u ->
+            vPrefer env (evalE t) (evalE u)
+        Merge x y ma ->
+            case (evalE x, evalE y, fmap evalE ma) of
+                (VRecordLit m, VInject _ k mt, _)
+                    | Just f <- Map.lookup k m -> maybe f (vApp f) mt
+                    | otherwise                -> error errorMsg
+                (x, y, ma) -> VMerge x y ma
+        ToMap x ma ->
+            case (evalE x, fmap evalE ma) of
+                (VRecordLit m, ma'@(Just _)) | null m ->
+                    VListLit ma' (Sequence.empty)
+                (VRecordLit m, _) ->
+                    let entry (k, v) =
+                            VRecordLit
+                                (Map.fromList
+                                    [ ("mapKey", VTextLit $ VChunks [] k)
+                                    , ("mapValue", v)
+                                    ]
+                                )
+
+                        s = (Sequence.fromList . map entry . Map.toList) m
+
+                    in  VListLit Nothing s
+                (x, ma) ->
+                    VToMap x ma
+        Field t k ->
+            vField (evalE t) k
+        Project t (Left ks) ->
+            vProjectByFields env (evalE t) (Dhall.Set.sort ks)
+        Project t (Right e) ->
+            case evalE e of
+                VRecord kts ->
+                    evalE (Project t (Left (Dhall.Set.fromSet (Map.keysSet kts))))
+                e' ->
+                    VProject (evalE t) (Right e')
+        Assert t ->
+            VAssert (evalE t)
+        Equivalent t u ->
+            VEquivalent (evalE t) (evalE u)
+        Note _ e ->
+            evalE e
+        ImportAlt t _ ->
+            evalE t
+        Embed a ->
+            VEmbed a
+  where
     evalE :: Expr Void a -> Val a
     evalE = eval env
```

Comment:
How about getting rid of this helper?

`eval env` is only 3 characters longer but a full identifier less that one has to keep it mind.